### PR TITLE
Add checks when the [] operator was used at risk

### LINKED ISSUE #5848: Implement checks for image()[]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Other

### Describe the bug

Currently we are using `image()[]` to generate new or edited images on the fly in `agg_image.cpp`. However, we need checks for this since it could lead to crashes if it tries to access an array element out of bounds.

Initial discussion here with example of such a case here:
https://github.com/ihhub/fheroes2/pull/5836#discussion_r973714517

### Save file

Not needed.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by bjauny**
Fixes #5848 

I added checks when it seemed needed, the rest was already in bounds.

### CODE REVIEW

File: src/fheroes2/agg/agg_image.cpp [hunk 1]
```diff
@@ -99,9 +99,11 @@ namespace
 
     void replaceTranformPixel( fheroes2::Image & image, const int32_t position, const uint8_t value )
     {
-        if ( image.transform()[position] != 0 ) {
-            image.transform()[position] = 0;
-            image.image()[position] = value;
+        if ( position < ( image.width() * image.height() ) ) {
+            if ( image.transform()[position] != 0 ) {
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:103]: Please combine these 2 conditions under one if.

File: src/fheroes2/agg/agg_image.cpp [hunk 2]
```diff
@@ -424,7 +426,10 @@ namespace fheroes2
                     Sprite & out = _icnVsSprite[id][i];
                     out = GetICN( ICN::EVIL_CAMPAIGN_BUTTONS, 0 + i );
                     // clean the button.
-                    Fill( out, 13 + 2 * i, 3 + 2 * i, 129 - 2 * i, 16, out.image()[13 - 7 * i + ( 5 + i ) * ( 145 - ( 4 * i ) )] );
+                    int32_t colorId = 13 - 7 * i + ( 5 + i ) * ( 145 - ( 4 * i ) );
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:429]: This is not color ID but a pixel position. Also this variable can be `const`.

File: src/fheroes2/agg/agg_image.cpp [hunk 3]
```diff
@@ -452,7 +457,10 @@ namespace fheroes2
                     Sprite & out = _icnVsSprite[id][i];
                     out = GetICN( ICN::X_CMPBTN, 0 + i );
                     // clean the button.
-                    Fill( out, 4, 3 + i, 132 - i, 16, out.image()[5 * 132] );
+                    int32_t colorId = 5 * 132;
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:460]: This is not color ID but a pixel position. Also this variable can be `const`.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @bjauny , I put several comments here. Could you please take a look?

File: src/fheroes2/agg/agg_image.cpp (hunk 1 above)
Comment (bjauny) [reply src/fheroes2/agg/agg_image.cpp]: Done.

File: src/fheroes2/agg/agg_image.cpp (hunk 2 above)
Comment (bjauny) [reply src/fheroes2/agg/agg_image.cpp]: Done.

File: src/fheroes2/agg/agg_image.cpp (hunk 3 above)
Comment (bjauny) [reply src/fheroes2/agg/agg_image.cpp]: Done.

**Comment by ihhub** [comment]
@bjauny , thank you so much for this fix!

**Pull request merged**